app/routes/thematic/geoencoder.py

```diff
@@ -0,0 +1,216 @@
+import re
+from typing import Any, Dict, List, Optional
+
+from fastapi import APIRouter, HTTPException, Query
+from unidecode import unidecode
+
+from app.crud.versions import get_version, get_version_names
+from app.errors import RecordNotFoundError
+from app.models.pydantic.responses import Response
+from app.routes import VERSION_REGEX
+from app.routes.datasets.queries import _query_dataset_json
+
+router = APIRouter()
+
+
+@router.get(
+    "/geoencode",
+    tags=["Geoencoder"],
+    status_code=200,
+)
+async def geoencode(
+    *,
+    admin_source: str = Query(
+        "GADM", description="The source of administrative boundaries to use."
+    ),
+    admin_version: str = Query(
+        ...,
+        description="Version of the administrative boundaries dataset to use.",
+    ),
+    country: str = Query(
+        description="Name of the country to match.",
+    ),
+    region: Optional[str] = Query(
+        None,
+        description="Name of the region to match.",
+    ),
+    subregion: Optional[str] = Query(
+        None,
+        description="Name of the subregion to match.",
+    ),
+    normalize_search: bool = Query(
+        True,
+        description="Whether or not to perform a case- and accent-insensitive search.",
+    ),
+):
+    """Look up administrative boundary IDs matching a specified country name
+    (and region name and subregion names, if specified).
+    """
+    admin_source_to_dataset: Dict[str, str] = {"GADM": "gadm_administrative_boundaries"}
+
+    try:
+        dataset: str = admin_source_to_dataset[admin_source.upper()]
+    except KeyError:
+        raise HTTPException(
+            status_code=400,
+            detail=(
+                "Invalid admin boundary source. Valid sources:"
+                f" {[source for source in admin_source_to_dataset.keys()]}"
+            ),
+        )
+
+    version_str: str = "v" + str(admin_version).lstrip("v")
+
+    await version_is_valid(dataset, version_str)
+
+    names: List[str | None] = sanitize_names(
+        normalize_search, country, region, subregion
+    )
+
+    adm_level: int = determine_admin_level(*names)
+
+    sql: str = _admin_boundary_lookup_sql(
+        adm_level, normalize_search, admin_source, *names
+    )
+
+    json_data: List[Dict[str, Any]] = await _query_dataset_json(
+        dataset, version_str, sql, None
+    )
+
+    return Response(
+        data={
+            "adminSource": admin_source,
+            "adminVersion": admin_version,
+            "matches": [
+                {
+                    "country": {
+                        "id": match["gid_0"].rsplit("_")[0],
```

Comment:
A function for getting the ID here would be easier to read

Reply:
Done!